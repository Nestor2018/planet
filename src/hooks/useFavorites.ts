import { useState, useEffect } from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";

const FAVORITES_KEY = "favorite_planets8";

interface Planet {
  id: string;
  [key: string]: any;
}

const useFavorites = () => {
  /**
   * State to store the list of favorite planets.
   */
  const [favorites, setFavorites] = useState<Planet[]>([]);

  useEffect(() => {
    /**
     * Loads the list of favorite planets from AsyncStorage.
     */
    const loadFavorites = async () => {
      const storedFavorites = await AsyncStorage.getItem(FAVORITES_KEY);
      if (storedFavorites) {
        setFavorites(JSON.parse(storedFavorites));
      }
    };
    loadFavorites();
  }, [favorites]);

  /**
   * Toggles the favorite status of a planet.
   * If the planet is already a favorite, it will be removed from the list.
   * If the planet is not a favorite, it will be added to the list.
   *
   * @param {Planet} planet - The planet to toggle in the favorites list.
   */
  const toggleFavorite = async (planet: Planet) => {
    const isFavorite = favorites.some(
      (favorite: Planet) => favorite.id === planet.id,
    );

    const updatedFavorites = isFavorite
      ? favorites.filter((favorite: Planet) => favorite.id !== planet.id)
      : [...favorites, planet];

    setFavorites(updatedFavorites);
    await AsyncStorage.setItem(FAVORITES_KEY, JSON.stringify(updatedFavorites));
  };

  return { favorites, toggleFavorite };
};

export default useFavorites;
