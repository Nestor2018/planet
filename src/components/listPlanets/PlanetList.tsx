import React from "react";
import { FlatList, TouchableOpacity, Text, StyleSheet } from "react-native";
import { StackNavigationProp } from "@react-navigation/stack";

import useFavorites from "../../hooks/useFavorites";
import colors from "../../utils/colors";
import { useNavigation } from "@react-navigation/native";
import Planet from "./Planet";
import { RootStackParamList } from "../../navigation/AppNavigator";
import { PlanetProps } from "../../interfaces/planet.interface";

const PlanetList: React.FC<PlanetProps> = ({ planets }) => {
  const navigation = useNavigation<StackNavigationProp<RootStackParamList>>();
  const { favorites } = useFavorites();

  return (
    <FlatList
      testID="planet-list"
      data={planets}
      keyExtractor={(item) => item.id}
      renderItem={({ item }) => (
        <TouchableOpacity
          onPress={() =>
            navigation.navigate("PlanetDetails", { planetId: item.rel })
          }
        >
          <Planet
            testID={`planet-${item.englishName}`}
            planet={item}
            isFavorite={favorites.some(
              (favorite: { id: string }) => favorite.id === item.id,
            )}
          />
        </TouchableOpacity>
      )}
      ListEmptyComponent={
        <Text style={styles.text}>No planets available.</Text>
      }
    />
  );
};

const styles = StyleSheet.create({
  text: {
    color: colors.white,
    fontSize: 16,
    marginTop: 16,
  },
});

export default PlanetList;
