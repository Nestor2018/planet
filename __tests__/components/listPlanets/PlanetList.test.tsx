import React from "react";
import { render, fireEvent } from "@testing-library/react-native";
import PlanetList from "../../../src/components/listPlanets/PlanetList";
import useFavorites from "../../../src/hooks/useFavorites";

jest.mock("../../../src/hooks/useFavorites.ts", () => ({
  __esModule: true,
  default: jest.fn(),
}));

jest.mock("../../../src/components/listPlanets/Planet", () => {
  const MockPlanet = ({ planet, isFavorite, ...props }) => (
    <view
      testID={`planet-${planet.id}`}
      planet={planet}
      isFavorite={isFavorite}
      {...props}
    />
  );
  MockPlanet.displayName = "Planet";
  return MockPlanet;
});

describe("PlanetList Component", () => {
  const mockPlanets = [
    { englishName: "Earth", id: "earth", rel: "earth-rel" },
    { englishName: "Mars", id: "mars", rel: "mars-rel" },
  ];

  const mockNavigate = jest.fn();

  it("renders a list of planets", () => {
    (useFavorites as jest.Mock).mockReturnValue({
      favorites: [{ id: "earth" }],
    });

    const { getByTestId } = render(<PlanetList planets={mockPlanets} />);

    const flatList = getByTestId("planet-list");
    expect(flatList).toBeTruthy();

    expect(getByTestId("planet-Earth")).toBeTruthy();
    expect(getByTestId("planet-Mars")).toBeTruthy();

    expect(getByTestId("planet-Earth").props.isFavorite).toBe(true);
    expect(getByTestId("planet-Mars").props.isFavorite).toBe(false);
  });

  it("renders empty message when no planets are provided", () => {
    (useFavorites as jest.Mock).mockReturnValue({
      favorites: [],
    });

    const { getByText, queryByTestId } = render(<PlanetList planets={[]} />);

    expect(queryByTestId("planet-Earth")).toBeNull();

    expect(getByText("No planets available.")).toBeTruthy();
  });

  it("updates favorite status based on useFavorites", () => {
    (useFavorites as jest.Mock).mockReturnValue({
      favorites: [{ id: "mars" }],
    });

    const { getByTestId } = render(<PlanetList planets={mockPlanets} />);

    expect(getByTestId("planet-Earth").props.isFavorite).toBe(false);
    expect(getByTestId("planet-Mars").props.isFavorite).toBe(true);
  });

  // 5. Snapshot
  it("matches snapshot with planets", () => {
    (useFavorites as jest.Mock).mockReturnValue({
      favorites: [{ id: "earth" }],
    });

    const tree = render(<PlanetList planets={mockPlanets} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
