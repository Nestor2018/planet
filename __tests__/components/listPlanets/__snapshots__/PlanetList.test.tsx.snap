// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanetList Component matches snapshot with planets 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <Text
      style={
        {
          "color": "#fff",
          "fontSize": 16,
          "marginTop": 16,
        }
      }
    >
      No planets available.
    </Text>
  }
  data={
    [
      {
        "englishName": "Earth",
        "id": "earth",
        "rel": "earth-rel",
      },
      {
        "englishName": "Mars",
        "id": "mars",
        "rel": "mars-rel",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={0.0001}
  stickyHeaderIndices={[]}
  testID="planet-list"
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <view
          isFavorite={true}
          planet={
            {
              "englishName": "Earth",
              "id": "earth",
              "rel": "earth-rel",
            }
          }
          testID="planet-Earth"
        />
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <view
          isFavorite={false}
          planet={
            {
              "englishName": "Mars",
              "id": "mars",
              "rel": "mars-rel",
            }
          }
          testID="planet-Mars"
        />
      </View>
    </View>
  </View>
</RCTScrollView>
`;
